core/src/main/java/io/grpc/internal/DnsNameResolver.java

```diff
@@ -130,18 +123,16 @@ public final synchronized void refresh() {
       public void run() {
         InetAddress[] inetAddrs;
         Listener savedListener;
-        synchronized (DnsNameResolver.this) {
-          // If this task is started by refresh(), there might already be a scheduled task.
-          if (resolutionTask != null) {
-            resolutionTask.cancel(false);
-            resolutionTask = null;
-          }
-          if (shutdown) {
-            return;
-          }
-          savedListener = listener;
-          resolving = true;
+        // If this task is started by refresh(), there might already be a scheduled task.
```

Comment:
What happens if resolutionRunnable runs on a thread other than the original one it ran on.  I don't know if ChannelExecutor is guaranteed to be serial.